File: built-in plugins dsp/mb-fm-synth.cpp
/**
 * MB FM Synth
 * Category : instrument
 * Type     : fm
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : FM synthesis engine with 4 operators
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FM_SYNTH_H
#define MB_FM_SYNTH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFmSynth : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-fm-synth";
    static constexpr const char* PLUGIN_NAME    = "MB FM Synth";
    static constexpr const char* PLUGIN_TYPE    = "fm";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float modIndex = 0.5f;  // range [0, 1]
    float ratio = 0.5f;  // range [0, 1]
    float brightness = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbFmSynth() = default;
    ~MbFmSynth() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.modIndex = std::clamp(params.modIndex, 0f, 1f);
        params.ratio = std::clamp(params.ratio, 0f, 1f);
        params.brightness = std::clamp(params.brightness, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB FM Synth
        return input;
    }
};

#endif // MB_FM_SYNTH_H

File: built-in plugins dsp/mb-fm-synth.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbFmSynthPlugin: PluginDefinition = {
  id: "mb-fm-synth",
  slug: "mb-fm-synth",
  name: "MB FM Synth",
  category: "instrument",
  type: "fm",
  version: "1.0.0",
  description: "FM synthesis engine with 4 operators",
  author: "Max Booster",
  grade: "A",
  oscillators: [{ type: "sine", detune: 0, gain: 1.0 }],
  envelope: { attack: 0.01, decay: 0.5, sustain: 0.4, release: 0.3 },
  parameters: [
    {
      id: "modIndex",
      name: "Mod Index",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "ratio",
      name: "Ratio",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "brightness",
      name: "Brightness",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { modIndex: 0.5, ratio: 0.5, brightness: 0.5, volume: 0.8 },
};

export default MbFmSynthPlugin;
